lib/galaxy/util/__init__.py

```diff
@@ -236,34 +242,39 @@ def unique_id(KEY_SIZE=128):
     return md5(random_bits).hexdigest()
 
 
-def parse_xml(fname):
+def parse_xml(fname, strip_whitespace=True, remove_comments=True):
     """Returns a parsed xml tree"""
-    # handle deprecation warning for XMLParsing a file with DOCTYPE
-    class DoctypeSafeCallbackTarget(ElementTree.TreeBuilder):
-        def doctype(*args):
-            pass
-    tree = ElementTree.ElementTree()
+    parser = None
+    if remove_comments and LXML_AVAILABLE:
+        # If using stdlib etree comments are always removed,
+        # but lxml doesn't do this by default
+        parser = etree.XMLParser(remove_comments=remove_comments)
     try:
-        root = tree.parse(fname, parser=ElementTree.XMLParser(target=DoctypeSafeCallbackTarget()))
-        for elem in root.iter('*'):
-            if elem.text is not None:
-                elem.text = elem.text.strip()
-            if elem.tail is not None:
-                elem.tail = elem.tail.strip()
-    except ParseError:
+        # Restore ENOENT that would otherwise be an OSError in lxml
```

Comment:
Weird filesystem semantics make it so it's usually safer to try open, catch, and raise if opening fails (usually faster than checking for existence first too) - is the `OSError`'s `errno` attribute set incorrectly in the case of lxml?

Plus IOErrors are OSErrors in Python 3.

Reply:
> is the `OSError`'s `errno` attribute set incorrectly in the case of lxml?

indeed, and that leads to this fun undefined behavior where we catch `IOError`s but then don't do anything if errno is not `ENOENT` - https://github.com/mvdbeek/galaxy/blob/a253acc31bf8aaa903e8e5fdbe44b31b204a0393/lib/galaxy/tools/data_manager/manager.py#L71.

I'll add another `except IOError` and change the errno there if appropriate.

Reply:
`e.errno` is None for missing files. I've added the missing errno in https://github.com/galaxyproject/galaxy/pull/9759/commits/0a8634b2f70be23c3fae85b88d067844acc2b998, I think this should be good enough as a workaround.